### ISSUE TITLE
`vue/no-setup-props-destructure` is too restrictive after latest update

### ISSUE THREAD

**Issue by l3d00m** [vuejs/eslint-plugin-vue#2259]
**What rule do you want to change?** 

`vue/no-setup-props-destructure`

**Does this change cause the rule to produce more or fewer warnings?** 

fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**

Revert #2244 and extract those rule changes into a new rule.

The changes now cause rule violations for example code directly from the vue documentation (see https://github.com/vuejs/eslint-plugin-vue/pull/2244#issuecomment-1656936626). 

It now enforces more or less subjective defaults. I mostly agree with the defaults, but the rule is active in `vue3-essential`, which shouldn't happen for subjective defaults.

In addition to that, the name and documentation no longer matches with the behaviour.

**Please provide some example code that this change will affect:**

```vue
<script setup>
  const props = defineProps({ count: Number })
  const count = ref(props.count)
                         ^^^^^^ Failing
</script>
```

**What does the rule currently do for this code?**

The code is violating the rule, even though this is not a destructure, but a non reactive access.

**What will the rule do after it's changed?**

It shouldn't violate. This is a valid use case for setting an initial value, even though it might not be required to do so often. Still, it shouldn't be necessary to disable the rule for every line where this occurs.

Please consider making this a new rule, which could be placed in `recommended` instead. This has nothing to do with destructuring anymore, which is indeed critical and should fail.

**Comment by websmurf** [comment]
We are also running into issues with the last updated version, for example: 

```typescript
<script lang="ts">
export default defineComponent({
  props: {
    canSave: {
      type: Boolean,
      default: true,
    },
    showSave: {
      type: Boolean,
      default: true,
    },
  },
  setup (props) {
    const disabled = ref<boolean>(props.canSave === false)
                                  ^^ Failing
```

or 
```typescript
<script lang="ts">
export default defineComponent({
  props: {
    id: {
      type: String,
      required: true,
    },
    modelValue: {
      type: Boolean,
      required: true,
    },
    noLabel: {
      type: String,
      required: true,
    },
    yesLabel: {
      type: String,
      required: true,
    },
  },
  emits: ['update:modelValue'],
  setup (props, { emit }) {
    ...
    const options: RadioInputItem[] = ref([
                                      ^^ Failing
      { value: false, label: props.noLabel },
      { value: true, label: props.yesLabel },
    ])
```

**Comment by l3d00m** [comment]
> It would be nice if we could exclude this case via additional option.

Please, do not make this an additional option. As stated in the issue, the linting is completely unrelated to what `vue/no-setup-props-destructure` implies so it should be a new rule, rather than an option.

It might be a good idea to revert the #2244 as a quick patch release (and extract it into a proper new rule with the next major/minor release), since this issue seems to affect a lot of people.

**Comment by ota-meshi** [comment]
The `vue/no-setup-props-destructure` rule warns of statements that lose reactivity of props values.
https://eslint.vuejs.org/rules/no-setup-props-destructure.html#rule-details
So I think the current rule is working correctly.
You can turn off the rule if you intentionally want to lose props reactivity.
If people don't find this rule useful, I think they can consider removing it from the shareable configuration provided by this plugin.
Also, if the sharable configuration provided by this plugin is not useful for people, we may consider deprecating the sharable configuration provided by the plugin in the next version.

**Comment by jacekkarczmarczyk** [comment]
I'm not personally against the new behaviour, but the name of the rule doesn't match this behaviour

**Comment by l3d00m** [comment]
> The `vue/no-setup-props-destructure` rule warns of statements that lose reactivity of props values.

Sorry, but this is not true. Your links explicitly states that the rule "reports **the destructuring of props** passed to setup causing the value to lose reactivity".

And my complaint was specifically about the newly added access linting.


> You can turn off the rule if you intentionally want to lose props reactivity.

No, because this rule is incredibly useful! Erroring on **destructuring** is important for me. Just the new changes are not.


> Also, if the sharable configuration provided by this plugin is not useful for people, we may consider deprecating the sharable configuration provided by the plugin in the next version.

I don't understand where you got that from. The only fact so many people are complaining here is because probably everyone uses the shareable config because they are incredibly helpful. I really do appreciate them.

Please understand that I'm not coming from a hostile position. I ask you to reconsider splitting the rules. 

The original rule is important and it makes sense to have it as an error rule in `essential`. The new changes though are subjective and should probably just emit a warning. And they most certainly shouldn't be placed in `essential` as an error. As a user there's currently no way but to disable the rules for both use cases.

**Comment by kleinfreund** [comment]
While I totally understand the frustration this brings with it, the rule as it used to work is incomplete in my opinion. Sure, the name is misleading now, but _both_ destructuring of `props` _and_ initialization of new refs using `props` via property access break in the same way:

```js
const newRef1 = ref(props.test)

const { test } = props
const newRef2 = ref(test)
```

Neither of the created refs are reactive based on `props.test`.

There is a simple solution though: [toRef](https://vuejs.org/api/reactivity-utilities.html#toref)

```js
const newRef3 = toRef(() => props.test)
```

So, if (and multiple people point this out here) the risk of unintentional loss of reactivity through destructuring of `props` is real and should lead to an error, the same should be true for other ways the same issue can occur. The changed rule addresses this.

**Comment by daniluk4000** [comment]
> There is a simple solution though: [toRef](https://vuejs.org/api/reactivity-utilities.html#toref)

toRef will create readonly ref from my prop, which is not the case. I'll share the example.

```typescript
const localValue = ref(props.modelValue);
const { modelValue } = props;

const handleValueUpdate = (value: string) => {
    emit('update:modelValue', value);
}

 // Watch and set actual value here
watch(() => props.modelValue, () => localValue.value = props.modelValue);

// Oops! Value remains old and will never change since we used destructure
watch(() => modelValue, () => localValue.value = modelValue);
</script>
```

As you can see in this example, destructuring of props would break my watch as well as anything else. Smart use of top-level assigment would not lead to this thought. What will be your solution for this case? 

Also, let's take a look on OAPI realization.

```vue
<script lang="ts">
import { defineComponent, ref } from 'vue';

export default defineComponent({
     props: {
          modelValue: {
              type: String,
              required: true,
          },
     },
     data() {
          return {
               localValue: this.modelValue,
          },
     },
     watch: {
         modelValue() {
              this.localValue = this.modelValue;
         },
     },
     methods: {
          handleValueUpdate(value: string) {
               this.$emit('update:modelValue', value);
          },
     },
});
</script>
```

This is one of classic cases when we need to set default value for our **changeable** data.

**Comment by kleinfreund** [comment]
> toRef will create readonly ref from my prop

It's intentional that props aren't allowed to be mutated, hence `toRef(() => props.test)` won't allow you to modify `props.test`.

---

Looking at this more, it seems to at least be hard to replicate some very common uses cases _while leaving the rule active_:

```js
const newRef = ref(props.test)
```

This results in an error as many people pointed out.

But how can one declare that they're fine with losing the reactivity in this example?

```js
const newRef = toRef(props.test)
```

This works, but you can't mutate `newRef`. What if you want a local piece of reactive state that you can mutate that's initialized with a prop value?

```js
const propValue = toValue(props.test)
const newRef = ref(propValue)
```

Unfortunately, the call to `toValue(props.test)` produces the rule error.

**Comment by davidmyersdev** [comment]
I'd like to point out that [vue/no-ref-object-destructure](https://eslint.vuejs.org/rules/no-ref-object-destructure.html) actually behaves like this rules does _after_ these changes. In fact, I based my PR on that rule after I noticed the disparity between them. That said, I also think there is confusion here with the meaning of _destructuring_ in this context. Take the following example:

```ts
const foo = { bar: 'baz' }
const { bar } = foo
```

It could just as easily be rewritten as this:

```ts
const foo = { bar: 'baz' }
const bar = foo.bar
```

No, the latter example is not a _destructuring assignment_, but it accomplishes the same thing. It destructures an object's property accesses into distinct references to the underlying objects or values, which breaks the ability to track said accesses (hence why it breaks reactivity). Preventing the _accidental_ breakage of reactivity is likely why you're using this rule, otherwise disabling it is a simple option here.

That said, yes, it makes sense to support cases where you're making a conscious decision to not keep the reactivity of props (or at least handle it differently), so let's dig into what you can do right now without a change to this rule.

> But how can one declare that they're fine with losing the reactivity in this example?
> 
> ```ts
> const newRef = toRef(props.test)
> ```

While some have pointed out the `toRef` example, the main complaint is that using it will cause you to lose the ability to update the underly `props` value. I'd like to recommend you don't do that, as the Vue docs clearly state that props flow [one way](https://vuejs.org/guide/components/props.html#one-way-data-flow), but I know that's not always practical in a pre-existing codebase. Instead, I'd recommend you look at [one of the overloads of `toRef`](https://vuejs.org/api/reactivity-utilities.html#toref) that allows you to maintain that 2-way data flow:

```ts
// Specify the prop as the second argument
const newRef = toRef(props, 'test')
```

**Comment by ota-meshi** [comment]
> I'm not personally against the new behaviour, but the name of the rule doesn't match this behaviour.

Yeah. I agree that the rule name is confusing. When the rule was implemented, I thought destructuring was the cause of losing reactivity, but it wasn't the only one.
We may need to change to the new rule name. Can you suggest a better rule name?

**Comment by ddenev** [comment]
> I'd like to point out that [vue/no-ref-object-destructure](https://eslint.vuejs.org/rules/no-ref-object-destructure.html) actually behaves like this rules does _after_ these changes. In fact, I based my PR on that rule after I noticed the disparity between them. That said, I also think there is confusion here with the meaning of _destructuring_ in this context. Take the following example:
> 
> ```ts
> const foo = { bar: 'baz' }
> const { bar } = foo
> ```
> 
> It could just as easily be rewritten as this:
> 
> ```ts
> const foo = { bar: 'baz' }
> const bar = foo.bar
> ```
> 
> No, the latter example is not a _destructuring assignment_, but it accomplishes the same thing. It destructures an object's property accesses into distinct references to the underlying objects or values, which breaks the ability to track said accesses (hence why it breaks reactivity). In fact, if you [look at the tests from before I modified the rule](https://github.com/vuejs/eslint-plugin-vue/blob/e9ec8992377f2bed6ece5abe04576888bac8a9c6/tests/lib/rules/no-setup-props-destructure.js#L373), they explicitly test that the latter example produces an error. Preventing the _accidental_ breakage of reactivity is likely why you're using this rule, otherwise disabling it is a simple option here.
> 
> That said, yes, it makes sense to support cases where you're making a conscious decision to not keep the reactivity of props (or at least handle it differently), so let's dig into what you can do right now without a change to this rule.
> 
> > But how can one declare that they're fine with losing the reactivity in this example?
> > ```ts
> > const newRef = toRef(props.test)
> > ```
> 
> While some have pointed out the `toRef` example, the main complaint is that using it will cause you to lose the ability to update the underly `props` value. I'd like to recommend you don't do that, as the Vue docs clearly state that props flow [one way](https://vuejs.org/guide/components/props.html#one-way-data-flow), but I know that's not always practical in a pre-existing codebase. Instead, I'd recommend you look at [one of the overloads of `toRef`](https://vuejs.org/api/reactivity-utilities.html#toref) that allows you to maintain that 2-way data flow:
> 
> ```ts
> // Specify the prop as the second argument
> const newRef = toRef(props, 'test')
> ```
> 
> My goal in making this change was not to ruin anyone's day. It was actually quite the opposite. As someone who loves Vue, I like to do my part to teach others about how it works, and a big part of that is the reactivity system. When beginners come to Vue and they don't fully understand how it works or they come from frameworks like React (where things are automatically reactive, albeit at the cost of performance), it's not difficult for them to break reactivity. When that happens, especially in a professional setting, it can be really discouraging for them. The goal here is to help catch errors like these and teach others in the process so that they can feel empowered to work in a Vue codebase with less risk.

That is a very good explanation of the case and thank you very much, @davidmyersdev, for that. Now it is clearer to me the reasoning behind this change and it makes more sense to cover the "losing reactivity" case. It's only the name of the rule that "gets in the way" but that's easily solvable.

What I was missing previously was how to rewrite my code so that I still can init my local state from props, not caring about reactivity, and not being bombarded by lint about these errors.
With your examples above this is now solved.

Therefore, one additional suggestion - what you specified above (the 2 approaches to initializing a local ref from props) is a VERY good candidate to be incorporated into the official Vue docs, to replace the current part about that topic. If you do that, IMHO the circle will be complete :)

**Comment by ddenev** [comment]
@davidmyersdev - one problem though :)

I just tested and this still fails with the new rule:
```vue
const newRef = toRef(props.test)
```

I guess you'll need to update the rule to ease on that case, right?

**Comment by jacekkarczmarczyk** [comment]
> We may need to change to the new rule name

I think this whole discussion suggest that this rule's name should stay, just the new behaviour should be reverted, and there should be a new rule which would forbid non-reactive assignments (how about `vue/no-non-reactive-assignments` as a name?) like `const { foo } = props` or `const foo = ref(props.foo)`.
Would it be difficult to detect all such assignments, not only related to props?
```js
const foo = someComputed.value;
const bar = anyReactive.someProp;
// etc
```
Of course the assignment should be outside of some effect to be reported, not sure how hard/easy it is to implement

**Comment by ota-meshi** [comment]
I have no plans to create new rule. As @kleinfreund wrote, the following two statements both lose reactivity.

```js
const newRef1 = ref(props.test)

const { test } = props
const newRef2 = ref(test)
```

I don't see any reason to allow only `newRef1`.

**Comment by richardsimko** [comment]
But then what is the canonical way of loading a prop as a default value in a ref or reactive object if I don't care about reactivity in the prop? Simply sprinkle the code with lint ignores?

```ts
  const state = reactive<{
    foo: string;
    bar: number;
  }>({
    // This works perfectly fine for loading a default value but will error with the new rule
    foo: props.modelValue.foo,
    isUnlocked: someOtherDefault,
  });
```

**Comment by ota-meshi** [comment]
If you intentionally want to lose reactivity, you should either turn off the rule or write a suppress comment.

**Comment by jacekkarczmarczyk** [comment]
> If you intentionally want to lose reactivity, you should either turn off the rule or write a suppress comment.

I'd agree with that if the name of the rule was different, but it is what it is and now it's misleading. And changing the name of the rule would be a breaking change.

**Comment by FloEdelmann** [comment]
See #2222 for a similar rule renaming.

**Comment by AndreiSoroka** [comment]
> I already agree that the name of the rule is confusing. We have to rename the rule. It will not be a breaking change. Copy the `vue/no-setup-props-destructure` rule to create a rule with a new name and deprecate the `vue/no-setup-props-destructure` rule. Remove deprecated rules in major version.

But `vue/no-setup-props-destructure` is a clear and understandable rule.
No need to merge with new logic, better to create the new rule.
P.s. Smaller rules are easier to manage, because easier to turn off
 ```json5
"rules": {
   "vue/new-rule": "off " // Good,
   // "vue/no-setup-props-destructure": "on" // from default
}
```
 but not 
```json5
"rules": {
   "vue/new-rule": ["error", null, { "subrule_to_turn_off_only_new_logic": false}] // Bad
}
```

**Comment by ota-meshi** [comment]
I think we need to make the following changes.

- Rename that rule. (Minor versions have rules with both names. Mark `no-setup-props-destructure` as deprecated.)
- Remove that rule from shareable configurations.

Can someone please suggest a new rule name?

---

I have no plans to revert the `vue/no-setup-props-destructure` rule. because that rule will soon be deprecated.

**Comment by kleinfreund** [comment]
> Can someone please suggest a new rule name?

Perhaps `vue/no-setup-props-reactivity-loss`.

This covers the new behavior better.

However, I do think that this is quite the subjective rule. I reckon that far more people have code like `const newRef = ref(props.test)` than code like `const { test } = defineProps(['test'])`. For this group, the rule will likely have lost its value as there isn't currently a good way to both keep the rule active and not always disable it whenever you try to initialize a new ref with a prop as its initial value.

Personally I think this change should’ve shipped as a breaking change so people had a clear point at which they could decide whether the rule has value for them.

Also, just to make this clear, I do think the presets this plugin provides are _very_ useful and I definitely think it should continue shipping them.

**Comment by AndreiSoroka** [comment]
> because that rule will soon be deprecated
Could you please explain why? Why need to expand the rule, but not create new rules?

Like this
```json5
{
  "vue/no-setup-props-destructure": "on",
  "vue/no-reactivity-loss-on-props-ref": "off"
}
```

P.s. From Chat GPT variants
- vue/no-reactivity-loss-on-props-ref
- vue/avoid-props-ref-reativity-loss
- vue/no-props-to-ref-directly
- vue/prevent-ref-from-props
- vue/props-direct-ref-disallowed
- vue/no-direct-props-ref

**Comment by richardsimko** [comment]
> If you intentionally want to lose reactivity, you should either turn off the rule or write a suppress comment.

So you mean that props as defaults for an mutable state is an invalid approach? Because I think I've seen it recommended in the Vue docs, although I may be wrong.

To me disabling a default rule implies I'm doing something that goes against best practices. If assigning a prop value as a default value is against best practices, then what is the best practice? How would you solve the same case without disabling this rule?

**Comment by jacekkarczmarczyk** [comment]
@richardsimko `defineModel`, `useVModel` from `vue-use` etc

**Comment by cailloumajor** [comment]
> @richardsimko `defineModel`, `useVModel` from `vue-use` etc

@jacekkarczmarczyk I don't think this answers @richardsimko need (that is, correct me if I'm wrong, to have a one-time one-way binding from parent to component). Your answer, if I am not mistaken, is about two-way binding.

**Comment by lianee** [comment]
I'm surprised nobody actually mentionned the fact that this rule also triggers when props "value" are passed to a function.
What is the point of erroring on a simple `console.log(props.foo`), and what is the use to pass reactive values to axios calls?
Also, I think that basic level of checking rules should not error for any example in Vue site. Next levels could be more opinionated, but the basic ones should comply with official Vue recipes.
For now, I totally deactivated this rule, I won't refactor perfectly legal code for this subjective rule.

**Comment by ota-meshi** [comment]
You can turn off the rule if you intentionally want to lose reactivity.

**Comment by ota-meshi** [comment]
Thanks to those who suggested new rule names.
I have already commented how to modify this plugin to solve this issue.
If you don't like it, you can fork this project.
I don't want to receive any more comments, so I'm locking this issue.

**Comment by l3d00m** [comment]
@ota-meshi How would you address usages of `console.log(props.foo)` or calls to external, non reactive libraries (like axios)? 

I'd say that disabling the rule for every usage is not feasible. And recommending to disable it globally shouldn't be the goal of a eslint plugin.

**Comment by FloEdelmann** [comment]
The rule has now been renamed to [`vue/no-setup-props-reactivity-loss`](https://eslint.vuejs.org/rules/no-setup-props-reactivity-loss.html) and removed from any preset configs.  
See https://github.com/vuejs/eslint-plugin-vue/releases/tag/v9.17.0.
